### PR TITLE
rustdoc: Unify macro intra-doc link resolution with type and value resolution

### LINKED ISSUE #81633: private macros break intra-doc links
```rust
mod bar {
    macro_rules! str {() => {}}
}

pub mod foo {
    use std::path::{Path, PathBuf};

    pub struct Foo(PathBuf);
    impl std::ops::Deref for Foo {
        type Target = Path;
        fn deref(&self) -> &Path { &self.0 }
    }
}
```

In the documentation of `Foo::to_str()` inherited from the deref impl, the documentation contains a link to ``[`&str`](str)`` that fails due to the presence of a `str!` macro in another module, even though that macro isn't accessible. (Even if it was, it shouldn't be affecting the resolution of an intra-doc link in the standard library.)

### Meta
`rustc --version --verbose`:
```
rustc 1.51.0-nightly (0e63af5da 2021-01-31)
binary: rustc
commit-hash: 0e63af5da3400ace48a0345117980473fd21ad73
commit-date: 2021-01-31
host: x86_64-unknown-linux-gnu
release: 1.51.0-nightly
LLVM version: 11.0.1
```

### PR DESCRIPTION
**Pull request by jyn514**
Fixes https://github.com/rust-lang/rust/issues/81633 and also cleans up quite a lot of code.

r? @Manishearth

### CODE REVIEW

**Comment by jyn514** [comment]
Oh, this can probably remove the `all_macros` field on the resolver too.

**Comment by jyn514** [comment]
Hmm, it looks like `resolve_str_path_error` doesn't consider `#[macro_export]` ... @petrochenkov would it be easy to make it do that or should I revert all the rustdoc changes and add back the call to `resolve_macro_path`?

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -396,29 +396,14 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
         path_str: &'a str,
         module_id: DefId,
     ) -> Result<Res, ResolutionFailure<'a>> {
-        let path = ast::Path::from_ident(Ident::from_str(path_str));
         self.cx.enter_resolver(|resolver| {
-            // FIXME(jynelson): does this really need 3 separate lookups?
```
Comment (Manishearth) [inline src/librustdoc/passes/collect_intra_doc_links.rs:401]: turns out, no :laughing:

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -396,29 +396,14 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
-            if let Ok((Some(ext), res)) = resolver.resolve_macro_path(
-                &path,
-                None,
-                &ParentScope::module(resolver.graph_root(), resolver),
-                false,
-                false,
```
Comment (Manishearth) [inline src/librustdoc/passes/collect_intra_doc_links.rs:407]: perhaps we should also add tests that ensure that macros can still be linked to on both newer and older editions

**Review by Manishearth — APPROVED** [review]
sweet, i always disliked this setup

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 2 above)
Comment (Manishearth) [reply src/librustdoc/passes/collect_intra_doc_links.rs:407]: Ah, yeah, you noted a problem with `#[macro_export]` above

Comment (Manishearth) [reply src/librustdoc/passes/collect_intra_doc_links.rs:407]: Yeah this code was like this to handle `macro_export`/old-style macro resolution. Which still exists. But perhaps we can handle this a bit cleaner.

**Comment by petrochenkov** [comment]
>Hmm, it looks like `resolve_str_path_error` doesn't consider `#[macro_export]`

`macro_export` is irrelevant, `resolve_str_path_error` doesn't consider non-modularized `macro_rules` scopes in general.

```rust
{
    struct S;

    macro_rules! mac { () => {} }
    // `mac`'s scope starts here

    /// `mac` <- `resolve_str_path_error` won't see this
   struct Z;

    //`mac`'s scope ends here
}
```

As you can see `fn resolve_str_path_error` will construct the scope in which the path is resolved using `ParentScope::module`, which leaves the `macro_rules` component of `ParentScope` empty.
Ideally the current `macro_rules` scope should be specifically tracked and updated in the current `ParentScope` like it's done in `BuildReducedGraphVisitor`, for example.

So both using and not using `all_macros` is wrong, but what is more wrong depends on whether you like false positives or false negatives more.

**Comment by petrochenkov** [comment]
>Ideally the current `macro_rules` scope should be specifically tracked and updated in the current `ParentScope`

This will be very easy to do for local `macro_rules` after link resolution is moved to `early.rs`, BTW.

This will be hard to do for links inlined from other crates, because we don't even keep the relevant information in metadata (although we may start keeping it due to macros 2.0).

**Comment by JohnCSimon** [comment]
merge conflicts

**Comment by jyn514** [comment]
yes, I'm aware, that's why it's marked waiting-on-author.

**Comment by petrochenkov** [comment]
Any chance to get this landed with `all_macros` kept and being used as a fallback if regular resolution fails?
It would be very helpful for ["rustdoc: Stop cloning the resolver"](https://github.com/rust-lang/rust/issues/83761).

**Comment by jyn514** [comment]
@petrochenkov I don't know when I'll get a chance to work on this. Feel free to take over the PR and have someone from the rustdoc team review.

**Comment by jyn514** [comment]
Ok I take it back, I can do this much:
> Any chance to get this landed with all_macros kept and being used as a fallback if regular resolution fails?

Not sure how that's helpful for the resolver hackery, since it still requires access to the resolver in `resolve_macro`, but it's easy enough. I'm not planning to work on removing `all_macros`.

**Comment by petrochenkov** [comment]
>Not sure how that's helpful for the resolver hackery

It removes the use of `resolve_macro_path` and leaves `resolve_str_path_error` as the only used resolution method.

**Comment by jyn514** [comment]
> macro_export is irrelevant, resolve_str_path_error doesn't consider non-modularized macro_rules scopes in general.

I'm not sure in what context you could document a non-modularized macro. Rustdoc doesn't anything at function-scope, only at the module level. I don't understand why the macro in the MCVE (https://github.com/rust-lang/rust/pull/91427#issuecomment-984036356) isn't resolved according to the rules you explained.

**Comment by Manishearth** [comment]
I may also have time later to do some rebasing and patchups here if necessary, but not any time soon.

**Comment by jyn514** [comment]
@Manishearth this requires more than patchups, someone needs to investigate why the MCVE doesn't work and then figure out how to change the resolver internals to make it work with `resolve_str_path_error`.

**Comment by jyn514** [comment]
Not planning to look into this. Happy for someone else to pick it up.

**Pull request closed without merging**